File: DivideChocolate.js
/**
 * @param {number[]} sweetness
 * @param {number} k
 * @return {number}
 */
var maximizeSweetness = function (sweetness, k) {
  let check = (sugar) => {
    let sum = 0;
    let pieces = 0;
    for (let i = 0; i < sweetness.length; i++) {
      sum += sweetness[i];
      if (sum >= sugar) {
        sum = 0;
        pieces++;
      }
    }
    return pieces >= k + 1;
  };
  let low = 1;
  let high = sweetness.reduce((acc, num) => (acc += num), 0) / (k + 1);
  while (low <= high) {
    let mid = Math.ceil((high + low) / 2);
    if (check(mid)) {
      low = mid + 1;
    } else {
      high = mid - 1;
    }
  }
  return high;
};

File: divideChocolate.py
class Solution:
    def maximizeSweetness(self, sweetness: List[int], k: int) -> int:
        result=0
        #greedy check that starts at an index and checks if the minimum is possible , as well as satisfying 
        #the minimum number of pieces 
        def checkSweetness(minSweetness):
            total = 0
            pieces=0
            for s in sweetness:
                total+=s
                if total >=minSweetness:
                    total = 0
                    pieces+=1
            return pieces >= k+1
        
        left = min(sweetness)
        right = sum(sweetness)
        #binary search on the solution space
        while left <= right:
            mid = (left + right )//2
            if checkSweetness(mid):
                result=mid
                left = mid+1
            else:
                right = mid -1
        return result
                


        